### PR TITLE
FIX: Do not keep old URL in history

### PR DESCRIPTION
**Pull request by udan11**
Clicking on an incomplete link to a topic (/t/ID or /t/SLUG) from
another post could replace current history entry or create two: one for
the incomplete URL and another one for the correct one. Going back was
either impossible or took the user to a redirect loop, redirected back
to /t/ID which redirected them again to /t/SLUG/ID.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/routes/topic-by-slug-or-id.js
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/routes/topic-by-slug-or-id.js:15]: This seems backwards to me... `replaceURL: true` should **prevent** the URL being kept in history. The default (`replaceURL: false`) should add a history entry. Do we have a bug deeper in the DiscourseURL module?

**Comment by eviltrout** [comment]
I want to confirm this as part of another bug I'm looking at it so please don't merge.

**Comment by eviltrout** [comment]
Looks like my bug is unrelated, so please ignore me :)

File: app/assets/javascripts/discourse/app/routes/topic-by-slug-or-id.js
Comment (udan11) [reply app/assets/javascripts/discourse/app/routes/topic-by-slug-or-id.js:15]: That is what I found weird about this too. I suspect it has something to do with the fact that the redirect happens while the URL is loading.

**Comment by SamSaffron** [comment]
I like the fix a lot, but agree with David here, it seems very odd to fix a bug by breaking a line :) 

If this is the only way to go about the fix I recommend a comment in the codebase that explains the oddity.

**Comment by ZogStriP** [comment]
Any news about this @udan11?

**Comment by udan11** [comment]
I updated the PR with a more descriptive commit message and a code comment. As soon as I get your approval, I will merge it.

**Review by eviltrout — APPROVED** [review]

**Pull request merged**
